plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -0,0 +1,988 @@
+mkUniswapValidator _  c (Pool lp a)   Remove      ctx = validateRemove c lp a ctx
+mkUniswapValidator _  c (Pool lp a)   Add         ctx = validateAdd c lp a ctx
+mkUniswapValidator _  _ _             _           _   = False
+
+validateLiquidityForging :: Uniswap -> TokenName -> ScriptContext -> Bool
+validateLiquidityForging us tn ctx = case [ i
+                                          | i <- txInfoInputs $ scriptContextTxInfo ctx
+                                          , let v = valueWithin i
+                                          , (coinValueOf v usC == 1) ||
+                                            (coinValueOf v lpC == 1)
+                                          ] of
+    [_]    -> True
+    [_, _] -> True
+    _      -> traceError "pool state forging without Uniswap input"
+  where
+    usC, lpC :: Coin
+    usC = usCoin us
+    lpC = Coin (ownCurrencySymbol ctx) tn
+
+uniswapInstance :: Uniswap -> Scripts.ScriptInstance Uniswapping
+uniswapInstance us = Scripts.validator @Uniswapping
+    ($$(PlutusTx.compile [|| mkUniswapValidator ||])
+        `PlutusTx.applyCode` PlutusTx.liftCode us
+        `PlutusTx.applyCode` PlutusTx.liftCode c)
+     $$(PlutusTx.compile [|| wrap ||])
+  where
+    c :: Coin
+    c = poolStateCoin us
+
+    wrap = Scripts.wrapValidator @UniswapDatum @UniswapAction
+
+uniswapScript :: Uniswap -> Validator
+uniswapScript = Scripts.validatorScript . uniswapInstance
+
+uniswapAddress :: Uniswap -> Ledger.Address
+uniswapAddress = Ledger.scriptAddress . uniswapScript
+
+uniswap :: CurrencySymbol -> Uniswap
+uniswap cs = Uniswap $ Coin cs uniswapTokenName
+
+liquidityPolicy :: Uniswap -> MonetaryPolicy
+liquidityPolicy us = mkMonetaryPolicyScript $
+    $$(PlutusTx.compile [|| \u t -> Scripts.wrapMonetaryPolicy (validateLiquidityForging u t) ||])
+        `PlutusTx.applyCode` PlutusTx.liftCode us
+        `PlutusTx.applyCode` PlutusTx.liftCode poolStateTokenName
+
+liquidityCurrency :: Uniswap -> CurrencySymbol
+liquidityCurrency = scriptCurrencySymbol . liquidityPolicy
+
+poolStateCoin :: Uniswap -> Coin
+poolStateCoin = flip Coin poolStateTokenName . liquidityCurrency
+
+-- | Gets the 'Coin' used to identity liquidity pools.
+poolStateCoinFromUniswapCurrency :: CurrencySymbol -- ^ The currency identifying the Uniswap instance.
+                                 -> Coin
+poolStateCoinFromUniswapCurrency = poolStateCoin . uniswap
+
+-- | Gets the liquidity token for a given liquidity pool.
+liquidityCoin :: CurrencySymbol -- ^ The currency identifying the Uniswap instance.
+              -> Coin           -- ^ One coin in the liquidity pair.
+              -> Coin           -- ^ The other coin in the liquidity pair.
+              -> Coin
+liquidityCoin cs coinA coinB = Coin (liquidityCurrency $ uniswap cs) $ lpTicker $ LiquidityPool coinA coinB
+
+-- | Paraneters for the @create@-endpoint, which creates a new liquidity pool.
+data CreateParams = CreateParams
```

Comment:
I wonder if we should split this module up more. Maybe we should encourage people to split all the off-chain stuff off into another module? In this case I wonder if we could split out some generic functions that just know about liquidity pools and operations on them too...